common/tasks/sequential_scheduler.go

```diff
@@ -0,0 +1,330 @@
+			targetWorkerNum := s.options.WorkerCount()
+			currentWorkerNum := len(s.workerShutdownCh)
+
+			if targetWorkerNum == currentWorkerNum {
+				continue
+			}
+
+			if targetWorkerNum > currentWorkerNum {
+				s.startWorkers(targetWorkerNum - currentWorkerNum)
+			} else {
+				s.stopWorkers(currentWorkerNum - targetWorkerNum)
+			}
+			s.logger.Info("Update worker pool size", tag.Key("worker-pool-size"), tag.Value(targetWorkerNum))
+		}
+	}
+}
+
+func (s *SequentialScheduler[T]) startWorkers(
+	count int,
+) {
+	for i := 0; i < count; i++ {
+		shutdownCh := make(chan struct{})
+		s.workerShutdownCh = append(s.workerShutdownCh, shutdownCh)
+
+		s.shutdownWG.Add(1)
+		go s.pollTaskQueue(shutdownCh)
+	}
+}
+
+func (s *SequentialScheduler[T]) stopWorkers(
+	count int,
+) {
+	shutdownChToClose := s.workerShutdownCh[:count]
+	s.workerShutdownCh = s.workerShutdownCh[count:]
+
+	for _, shutdownCh := range shutdownChToClose {
+		close(shutdownCh)
+	}
+}
+
+func (s *SequentialScheduler[T]) pollTaskQueue(workerShutdownCh <-chan struct{}) {
+	defer s.shutdownWG.Done()
+
+	for {
+		select {
+		case <-s.shutdownChan:
+			return
+		case <-workerShutdownCh:
+			return
+		case queue := <-s.queueChan:
+			s.processTaskQueue(queue, workerShutdownCh)
+		}
+	}
+}
+
+func (s *SequentialScheduler[T]) processTaskQueue(
+	queue SequentialTaskQueue[T],
+	workerShutdownCh <-chan struct{},
+) {
+	for {
+		select {
+		case <-s.shutdownChan:
+			s.drainTasks()
+			return
+		case <-workerShutdownCh:
+			// Put queue back to the queue channel
+			s.queueChan <- queue
+			return
+		default:
+			if !queue.IsEmpty() {
+				s.executeTask(queue)
+			} else {
+				deleted := s.queues.RemoveIf(queue.ID(), func(key interface{}, value interface{}) bool {
+					return value.(SequentialTaskQueue[T]).IsEmpty()
+				})
+				if deleted {
+					return
+				}
+				// if deletion failed, meaning that task queue is offered with new task
+				// continue execution
+			}
+		}
+	}
+}
+
+// TODO: change this function to process all available tasks in the queue.
+func (s *SequentialScheduler[T]) executeTask(queue SequentialTaskQueue[T]) {
+	task := queue.Remove()
+	operation := func() error {
+		if err := task.Execute(); err != nil {
+			return task.HandleErr(err)
+		}
+		return nil
+	}
+	isRetryable := func(err error) bool {
+		return !s.isStopped() && task.IsRetryableError(err)
+	}
+	if err := backoff.ThrottleRetry(operation, task.RetryPolicy(), isRetryable); err != nil {
+		if s.isStopped() {
+			task.Abort()
+			return
+		}
+
+		task.Nack(err)
+		return
+	}
+
+	task.Ack()
+}
+
+func (s *SequentialScheduler[T]) drainTasks() {
+LoopDrainQueue:
+	for {
+		select {
+		case queue := <-s.queueChan:
+			for !queue.IsEmpty() {
+				queue.Remove().Abort()
+			}
+			deleted := s.queues.RemoveIf(queue.ID(), func(key interface{}, value interface{}) bool {
+				return value.(SequentialTaskQueue[T]).IsEmpty()
+			})
+			if deleted {
+				return
+			}
```

Comment:
maybe I missed something. This just means  `queue` is successfully deleted from `s.queues` right, and we should continue to drain the next queue in `s.queueChan` ?